0004-post-tags.md

```diff
@@ -0,0 +1,324 @@
+
+### Protocol:
+According to https://www.w3.org/TR/activitystreams-vocabulary/#dfn-tags a general tag object exists in the ActivityPub protocol allowing for custom implementations so long as a type is specified.
+To keep things simple an initial community tag object should consist of the following:
+```json
+{
+  id: number,
+  url: string,
+  name: string
+}
+```
+
+Entirely unmoderated tags are not an option for lemmy as the moderation workload would be too much. Additionally users being able to type out tags themselves introduces splintering in the tag contents due to typos. A better solution is a curated list of tags users can attach to their posts. The list of tags can be maintained by both admins and moderators allowing for each community to tailor tags to their specific needs.
+Content for tags would be located in the respective root: `https://example.org/c/example/t/tag`
+
+The tag URL could then be utilized as a unique identifier for the tag, however doing so would restrict the ability to move the underlying tag url in the future. Instead an additional `id` field should be used. Using the object ID optional tag federation can also be achieved, allowing for communities across multiple instances to share content via tags (example: News tag shared across instances). This would also solve the issue of splintered communities across instances while not forcing it on the communities in question. For now tags will only be applicable to posts, however the general design allows for them to be attached to any kind of object later on, be that instance, community or user. The limited initial scope allows for easier modifications should any rough edges or missing features be discovered.
+
+To fill all needed use cases 3 tag flavors will be needed:
+
+- NSFW
+- Content Warning
+- Generic Tag
+
+Theoretically NSFW could be implemented using a preset "Content Warning" tag but seperating out this tag allows instances to better filter it out for moderation purposes (for example if no admin/moderator is willing to moderate NSFW content).
+Both "NSFW" and "Content Warning" tags should blur the post body by default. Additionally the `sensitive` post flag to `true` should either of these flavors be present in the post tags to ensure correct handling on other fediverse platforms.
+
+Example Tag Objects:
+
+**Generic Tag:**
+```json
+{
+    "id": 1,
+    "url": "https://example.org/c/community/t/tag",
+    "name": "Generic Tag"
+}
+```
+
+**NSFW Tag:**
+```json
+{
+    "flavor": "nsfw",
+    "id": 1,
+    "url": "https://example.org/c/adult_community/t/nsfw",
+    "name": "NSFW"
+}
+```
+
+**Spoiler Tag:**
+```json
+{
+    "flavor": "cw",
+    "id": 1,
+    "url": "https://example.org/c/story_community/t/spoiler",
+    "name": "Spoiler"
+}
+```
+
+**Remote Instance Spoiler Tag:**
+```json
+{
+    "flavor": "cw",
+    "id": 1,
+    "url": "https://example.org/c/example_community/tag@example_remote.org",
+    "name": "Spoiler"
+}
+```
```

Comment:
The "remote instance" example doesn't make sense, since the activitypub URL of the tag will stay the same regardless of which instance is looking at the post object.

The "flavor" attribute IMO doesn't make much sense and isn't really described in enough detail, it seems like instead it makes more sense to add a `hierarchy: "content-warning/triggers/gore"` or `inherits:` property to tags in the future if we want specific tags that imply generic tags.

Reply:
Agreed, the remote instance example was a symptom of me not really understanding activitypub (which I think is pretty obvious by the extended fixing needed in that section so far, thank you for that a lot)

Personal style preference but I would opt for the `inherits` field instead of the `hierarchy` one, reference the parent tag URL there once and have the server resolve the tag recursively if not already known instead of shipping the entire inheritance tree with every occurrence of the tag. Especially since only the immediate parent needs to be known to sort out the correct display behaviour of a tag. Leaving the field empty (or omitting it entirely, not sure which one would be best practice) would then mean the tag is a root tag that starts a new behaviour tree (for now only display behaviour).

Including this field later should be possible but would absolutely require a way for admins and moderators to change a tag's parent from the settings (something I think is a good idea anyway).
I've moved this specific part to the Future Possibilities Section and added a bit more detail to my thoughts. I also replaced the previous example tags with the tags present in the post example, it's small change but I think it elevates the readability.

See d582286ba980989272d375da00c686bcc99fdade for the changes

Reply:
The text before this code section still needs to be changed by removing the sentence "In order to share this news post across instances it also is tagged with the news tag of a remote instance".

Reply:
it might make sense to do what you say, but it's also a lot more complex. IMO tags shouldn't need resolution or complex whatever algorithms at all, the hierarchy should purely be a way to allow prefix-based filtering in a bit of a structured way. everything you need to know about the tag should be contained within the tag.